src/plugins/es_ui_shared/public/components/json_editor/use_json.ts

```diff
@@ -62,31 +62,38 @@ export const useJson = <T extends object = { [key: string]: any }>({
       setError(null);
     }
     return isValid;
-  };
+  }, [content]);
 
-  const formatContent = () => {
+  const formatContent = useCallback(() => {
     const isValid = validate();
     const data = isValid && content.trim() !== '' ? JSON.parse(content) : {};
     return data as T;
-  };
+  }, [validate, content]);
 
   useEffect(() => {
-    if (didMount.current) {
-      const isValid = isControlled ? undefined : validate();
-      onUpdate({
-        data: {
-          raw: content,
-          format: formatContent,
-        },
-        validate,
-        isValid,
-      });
-    } else {
-      didMount.current = true;
+    if (!isMounted.current) {
```

Comment:
I'm having trouble figuring out how we'd get into this state. Did you run into a situation where the component wasn't mounted when this function was called?

Reply:
Lol that's a great catch! 😊 I refactored a bit to quickly, I will put back how it was before. The intention was to not call the `onUpdate` on mount as... nothing has been updated yet.

Reply:
EDIT: Looking at the code it is correct. So like I said, the idea is not to call the `onUpdate` on component mount as there isn't any update.

In the next `useEffect` the component `isMounted.current` will be set to true so on the next update it will call the handler.

Reply:
Ah, I see. I think the use of the name `isMounted` is confusing to me, particularly because it defaults to `false`. The first time a hook is called is when the owner mounts, so by definition when I'm walking through a hook's "lifecycle" I begin with the mindset of it being mounted somewhere. To fix this, how about renaming the variable `isInitialCall` or `isFirstUpdate`? I think this area of the code also deserves a comment explaining your intentions to the reader.

Reply:
> particularly because it defaults to false

I must default to false. when the ref is created the component has not mounted yet. It only has mounted when a `useEffect` is called (every `useEffect` is called **in order of appearance** in the code. I learned that the hard way). 

So the convention I have put in place everywhere is the same: `isMounted` (because that is the React term for this and I prefer not to add another concept) defaults to `false` and is set to `true` **in the last** `useEffect` in the code. So any previous useEffect will be called when mounting but the ref will still be false for them. This is exactly what we want: have a ref in our useEffects telling us if the component has mounted or not, and have some conditional logic to prevent doing certain things "on component mount". As you commented somewhere, `useEffect` should have a single purpose and not one giant `useEffect` (which would probably not work anyway).

This is what I shared in my "Lesson learned" here https://github.com/elastic/kibana/pull/75796

Reply:
👍  Thanks for explaining. I think some comments in the code would help explain this to others (or even our future selves).